File: domain/src/main/java/org/openmedstack/eventstore/CommitAttempt.java
package org.openmedstack.eventstore;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;

public class CommitAttempt {
    private String _bucketId;
    private String _streamId;
    private int _streamRevision;
    private UUID _commitId;
    private int _commitSequence;
    private Instant _commitStamp;
    private HashMap<String, Object> _headers;
    private List<EventMessage> _events;

    public CommitAttempt(
            String bucketId,
            String streamId,
            int streamRevision,
            UUID commitId,
            int commitSequence,
            Instant commitStamp,
            HashMap<String, Object> headers,
            List<EventMessage> events) {
        _bucketId = bucketId;
        _streamId = streamId;
        _streamRevision = streamRevision;
        _commitId = commitId;
        _commitSequence = commitSequence;
        _commitStamp = commitStamp;
        _headers = headers == null ? new HashMap<>() : headers;
        _events = events == null ?
                new ArrayList<>() :
                events;
    }

    public String getBucketId() {
        return _bucketId;
    }

    public String getStreamId() {
        return _streamId;
    }

    public int getStreamRevision() {
        return _streamRevision;
    }

    public UUID getCommitId() {
        return _commitId;
    }

    public int getCommitSequence() {
        return _commitSequence;
    }

    public Instant getCommitStamp() {
        return _commitStamp;
    }

    public HashMap<String, Object> getHeaders() {
        return _headers;
    }

    public List<EventMessage> getEvents() {
        return _events;
    }
}

File: domain/src/main/kotlin/org/openmedstack/eventstore/CommitAttempt.kt
package org.openmedstack.eventstore

import java.time.Instant
import java.util.*

class CommitAttempt(
        private val _bucketId: String?,
        private val _streamId: String,
        private val _streamRevision: Int,
        private val _commitId: UUID,
        private val _commitSequence: Int,
        private val _commitStamp: Instant,
        private val _headers: HashMap<String, Any>,
        private val _events: MutableList<EventMessage>) {
    val bucketId: String?
        get() = _bucketId

    val streamId: String
        get() = _streamId

    val streamRevision: Int
        get() = _streamRevision

    val commitId: UUID
        get() = _commitId

    val commitSequence: Int
        get() = _commitSequence

    val commitStamp: Instant
        get() = _commitStamp

    val headers: HashMap<String, Any>
        get() = _headers

    val events: MutableList<EventMessage>
        get() = _events
}

class CommitImpl private constructor(
    private val _bucketId: String?,
    private val _streamId: String,
    private val _streamRevision: Int,
    private val _commitId: UUID,
    private val _commitSequence: Int,
    private val _commitStamp: Instant,
    private val _headers: HashMap<String, Any>,
    private val _events: MutableList<EventMessage>): Commit {
    override val bucketId: String?
        get() = _bucketId
    override val streamId: String
        get() = _streamId
    override val streamRevision: Int
        get() = _streamRevision
    override val commitId: UUID
        get() = _commitId
    override val commitSequence: Int
        get() = _commitSequence
    override val commitStamp: Instant
        get() = _commitStamp
    override val headers: HashMap<String, Any>
        get() = _headers
    override val events: MutableList<EventMessage>
        get() = _events
    override val checkpointToken: Long
        get() = 0

    companion object {
        fun create(attempt: CommitAttempt): Commit {
            return CommitImpl(
                attempt.bucketId,
                attempt.streamId,
                attempt.streamRevision,
                attempt.commitId,
                attempt.commitSequence,
                attempt.commitStamp,
                attempt.headers,
                attempt.events
            )
        }
    }
}
